src/sys.c

```diff
@@ -514,6 +517,35 @@ JL_DLLEXPORT JL_STREAM *jl_stdin_stream(void)  { return JL_STDIN; }
 JL_DLLEXPORT JL_STREAM *jl_stdout_stream(void) { return JL_STDOUT; }
 JL_DLLEXPORT JL_STREAM *jl_stderr_stream(void) { return JL_STDERR; }
 
+// terminal workarounds
+JL_DLLEXPORT int jl_getch(void) JL_NOTSAFEPOINT
+{
+#if defined(_OS_WINDOWS_)
+    // Windows has an actual `_getch()`, use that:
+    return _getch();
+#else
+    // On all other platforms, we do the POSIX terminal manipulation dance
+    char buf = 0;
+    struct termios old = {0};
+    fflush(stdout);
+    if(tcgetattr(0, &old) < 0)
+        jl_error("tcgetattr()");
+    old.c_lflag &= ~ICANON;
+    old.c_lflag &= ~ECHO;
+    old.c_cc[VMIN] = 1;
+    old.c_cc[VTIME] = 0;
+    if(tcsetattr(0, TCSANOW, &old) < 0)
+        jl_error("tcsetattr(TCSANOW)");
+    if(read(0, &buf, 1) < 0)
+        jl_error("read()");
+    old.c_lflag |= ICANON;
+    old.c_lflag |= ECHO;
+    if(tcsetattr(0, TCSADRAIN, &old) < 0)
+        jl_error("tcsetattr(TCSADRAIN)");
+    return buf;
```

Comment:
```suggestion
    char c;
    int r;
    struct termios old = {0};
    struct termios new;
    if (tcgetattr(0, &old) != 0)
        return -1;
    new = old;
    cfmakeraw(&new);
    if (tcsetattr(0, TCSADRAIN, &new) != 0)
        return -1;
    r = read(0, &c, 1);
    if (tcsetattr(0, TCSADRAIN, &old) != 0)
        return -1;
    return r == 1 ? c : -1;
```

Reply:
`cfmakeraw` is more convenient, though adds a lot of other options too. I think that is what we want, but it does disable things like ^C and ^S.

Reply:
Anything here (get, set, read) might get interrupted by a signal and requires handling EINTR as well